File: Erlang/caesar.erl
-module(caesar). 
-import(lists,[reverse/1]). 
-export([start/0,shiftLetter/2,encrypt/2,decrypt/2,solve/2]).

shiftLetter(Letter, NumShift)  ->
	GetLetterValue = fun([Character]) -> Character end,
	LetterValue = GetLetterValue(Letter),
	if
	    LetterValue >= 65 ->
	        if
	            LetterValue =< 90 ->
	                (((GetLetterValue(Letter) - 65 + NumShift) rem (26) + 26) rem 26) + 65;
	            true ->
	                Letter
            end;
	    true ->
	        Letter
	end.

encrypt(Sentence, NumShift) ->
    Shifted = lists:map(fun(Letter) -> shiftLetter([Letter], NumShift) end, string:to_upper(Sentence)),
	io:format("~s", [Shifted]).
	
decrypt(Sentence, NumShift) -> encrypt(Sentence, NumShift * -1).

solve(Sentence, MaxShift) -> 
   Numbers = lists:seq(0, MaxShift),
   lists:foreach(fun(MaxShift) ->
		io:fwrite("\nCaesar ~p: ", [MaxShift]),
		decrypt(Sentence, MaxShift * -1) end, reverse(Numbers)).

start() ->
	encrypt("Hello, world!", 3),
	io:fwrite("\n"),
	decrypt("jkraixan", 22),
	io:fwrite("\n"),
	solve("hal", 26).
	

File: Haskell/caesar.hs
import Data.Char (ord, chr, toUpper)

shiftLetter :: Char -> Int -> Char
shiftLetter letter numShift = chr $ 65 + (ord (toUpper letter) - 65 + numShift) `mod` 26

encrypt :: Int -> String -> String
encrypt numShift = map sentence
 where sentence letter
              | toUpper letter >= 'A' && toUpper letter <= 'Z' = shiftLetter letter numShift
              | otherwise = letter
              
decrypt :: Int -> String -> String
decrypt shift = encrypt (-shift)

solve :: String -> Int -> IO ()
solve sentence maxShift
 | maxShift > -1 = do
  putStrLn ("Caesar " ++ (show maxShift) ++ ": " ++ (encrypt maxShift sentence))
  solve sentence (maxShift - 1)

main :: IO()
main = do
  putStrLn(encrypt 3 "Hello, world!")
  putStrLn(decrypt 15 "iwpczhvxkxcv")
  solve "hal" 26 
